api/graphite.go

```diff
@@ -673,3 +673,143 @@ func getLocation(desc string) (*time.Location, error) {
 	}
 	return time.LoadLocation(desc)
 }
+
+func (s *Server) graphiteTagDetails(ctx *middleware.Context, request models.GraphiteTagDetails) {
+	tag := ctx.Params(":tag")
+	if len(tag) <= 0 {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, "not tag specified"))
+		return
+	}
+
+	tagValues, err := s.clusterTagDetails(ctx.Req.Context(), ctx.OrgId, tag, request.Filter, request.From)
+	if err != nil {
+		response.Write(ctx, response.WrapError(err))
+		return
+	}
+
+	resp := models.GraphiteTagDetailsResp{
+		Tag:    tag,
+		Values: make([]models.GraphiteTagDetailsValueResp, 0, len(tagValues)),
+	}
+
+	for k, v := range tagValues {
+		resp.Values = append(resp.Values, models.GraphiteTagDetailsValueResp{
+			Value: k,
+			Count: v,
+		})
+	}
+
+	response.Write(ctx, response.NewJson(200, resp, ""))
+}
+
+func (s *Server) clusterTagDetails(ctx context.Context, orgId int, tag, filter string, from int64) (map[string]uint64, error) {
+	result, err := s.MetricIndex.TagDetails(orgId, tag, filter, from)
+	if err != nil {
+		return nil, err
+	}
+
+	path := "/index/tags/" + tag
+	data := models.IndexTagDetails{OrgId: orgId, Tag: tag, Filter: filter, From: from}
+	resp := &models.IndexTagDetailsResp{}
+	responses, err := s.peerQuery(ctx, data, "clusterTagDetails", path, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, resp := range responses {
+		for k, v := range resp.(*models.IndexTagDetailsResp).Values {
+			result[k] = result[k] + v
+		}
+	}
+
+	return result, nil
+}
+
+func (s *Server) graphiteTagFindSeries(ctx *middleware.Context, request models.GraphiteTagFindSeries) {
+	series, err := s.clusterTagFindSeries(ctx.Req.Context(), ctx.OrgId, request.Expr, request.From)
+	if err != nil {
+		response.Write(ctx, response.WrapError(err))
+		return
+	}
+
+	response.Write(ctx, response.NewJson(200, series, ""))
+}
+
+func (s *Server) clusterTagFindSeries(ctx context.Context, orgId int, expressions []string, from int64) ([]string, error) {
+	seriesSet := make(map[string]struct{})
+
+	result, err := s.MetricIndex.FindByTag(orgId, expressions, from)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, series := range result {
+		seriesSet[series] = struct{}{}
+	}
+
+	data := models.IndexTagFindSeries{OrgId: orgId, Expressions: expressions, From: from}
+	resp := &models.IndexTagFindSeriesResp{}
+	responses, err := s.peerQuery(ctx, data, "clusterTagFindSeries", "/index/tags/findSeries", resp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, resp := range responses {
+		for _, series := range resp.(*models.IndexTagFindSeriesResp).Series {
+			seriesSet[series] = struct{}{}
+		}
+	}
+
+	series := make([]string, 0, len(seriesSet))
+	for s := range seriesSet {
+		series = append(series, s)
+	}
```

Comment:
For the record, I assume the intermediate set is to remove duplicates.
1) we may get the same metric from multiple instances
2) we may get the same metric from the same instance 

I think both of these could happen when people start publishing the same metric but under a different interval, and/or it goes to a different partition (FYI)

note that we do not have duplicate surpression in FindByTag. I think we should add it there too.

Reply:
What's the benefit of adding it to `FindByTag`? currently there is no way to call `FindByTag` without also going through the deduplication in `clusterTagFindSeries`. I guess one could argue that in the future such a case might get added, but until then this would just be a waste of cpu cycles or not?

Reply:
you're right. so until then, let's add a warning to the function (both implementation and index method) that it may return duplicates

Reply:
https://github.com/grafana/metrictank/pull/750/commits/2d9cf9ef11fe59d2ac0b90d10811be5d9c9dad03